### PR TITLE
Issue 3.0

### PR DESCRIPTION
**Pull request by madateamprojekt**
Not sure where the mergeconflicts come from.

### CODE REVIEW

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/base/impl/FeatureAttribute.java [hunk 1]
```diff
@@ -39,54 +40,148 @@
...
+		setTypeFromString(type);
+		this.unit = unit;
+		this.recursive = recursive;
+		this.configurable = configurable;
 	}
 
-	public boolean getRecursive() {
-		return recursive;
+	public boolean checkValue() {
+		if (type.toString().equals("LONG")) {
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/base/impl/FeatureAttribute.java]: Define final Strings for LONG, DOUBLE and BOOLEAN.

**Comment by tthuem** [comment]
Please rerun builds in target travis. Failing tests seem to be performance related and should go through next time.

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/base/impl/FeatureAttribute.java [hunk 2]
```diff
@@ -39,54 +40,148 @@
...
+		if (type.toString().equals("DOUBLE")) {
+			return Double.parseDouble(value);
+		}
+		if (type.toString().equals("BOOLEAN")) {
+			if (value.toLowerCase().equals("true")) {
+				return true;
+			}
+			if (value.toLowerCase().equals("false")) {
+				return false;
+			}
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/base/impl/FeatureAttribute.java]: just

return value.toLowerCase().equals("true");

instead

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java [hunk 3]
```diff
@@ -185,62 +185,14 @@ protected void addDescription(Document doc, IFeature feat, Element fnod) {
...
 	 * @param attributeList
 	 * @param listElementName
 	 * @return true if an attributeName is already in the list
 	 */
 	private boolean checkAttributeList(String attributeName, LinkedList<FeatureAttribute> attributeList) {
-		for (int i = 0; i < attributeList.size(); i++) {
-			if (attributeList.get(i).getName().toLowerCase().equals(attributeName)) {
+		attributeName = attributeName.toLowerCase();
+		for (final FeatureAttribute fa : attributeList) {
+			if (fa.getName().toLowerCase().equals(attributeName)) {
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java:195]: return fa.getName().toLowerCase().equals(attributeName);

instead

edit: i was wrong

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java [hunk 4]
```diff
@@ -454,57 +377,99 @@ private void parseAttribute(Element e, LinkedList<FeatureAttribute> attributeLis
 			LinkedList<FeatureAttributeInherited> inheritedList) throws UnsupportedModelException {
 
 		if (e.hasAttributes()) {
+
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java]: unnecessary line break

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java [hunk 5]
```diff
@@ -454,57 +377,99 @@ private void parseAttribute(Element e, LinkedList<FeatureAttribute> attributeLis
...
+				} else {
+					throwError("Unknown parameter", e);
 				}
+			}
 
-				if (isNameInAttributeList(recursiveList, attribute.getName().toLowerCase())) {
-					throwError("Too many parameters in this inherited attribute: " + attribute.toString(), e);
-				}
+			if (name.isEmpty()) {
+				throwError("This attribute has needs a name.", e);
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java]: ????????????

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java [hunk 6]
```diff
@@ -454,57 +377,99 @@ private void parseAttribute(Element e, LinkedList<FeatureAttribute> attributeLis
...
+	}
+
+	/**
+	 * @param name
+	 * @param value
+	 * @param inheritedList
+	 * @return 0 if attribute was added to recursive list 1 if value has wrong type 2 if attribute was not inherited
+	 */
+	private int checkRecursiveList(String name, String value, LinkedList<FeatureAttributeInherited> inheritedList) {
+		name = name.toLowerCase();
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java]: Define names for integers returned by this method.

**Review by PaulWestphal — CHANGES_REQUESTED** [review]

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java [hunk 7]
```diff
@@ -192,13 +192,31 @@ protected void addDescription(Document doc, IFeature feat, Element fnod) {
 	private boolean checkAttributeList(String attributeName, LinkedList<FeatureAttribute> attributeList) {
 		attributeName = attributeName.toLowerCase();
 		for (final FeatureAttribute fa : attributeList) {
-			if (fa.getName().toLowerCase().equals(attributeName)) {
-				return true;
-			}
+			return (fa.getName().toLowerCase().equals(attributeName));
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XmlFeatureModelFormat.java]: I was wrong about this, sorry. It now returns true or false immediately, please revert to the original.

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XMLFeatureModelTags.java [hunk 8]
```diff
@@ -76,5 +76,7 @@
 	static final String RECURSIVE = "recursive";
 	static final String CONFIGURABLE = "configurable";
 
-	static final String WRONG_TYPE = "wrong_type";
+	static final String FEATURE_INHERITED = "feaute inherited";
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/io/xml/XMLFeatureModelTags.java]: Consider using enums for this.

**Review by PaulWestphal — CHANGES_REQUESTED** [review]

**Review by PaulWestphal — APPROVED** [review]

**Pull request merged**